aten/src/ATen/native/cuda/Assert.cuh

```diff
@@ -0,0 +1,230 @@
+      *size_ptr = (size_t)(dst - start);
+
+      // fix alignment
+      ptrdiff_t offset = dst - start;
+      ptrdiff_t misalign = offset % C10_ASSERT_ARG_ALIGN_SIZE;
+      if (misalign > 0) {
+        dst += C10_ASSERT_ARG_ALIGN_SIZE - misalign;
+      }
+
+      // again check position after alignment correction
+      if (dst <= buffer_end) {
+        return dst;
+      }
+    }
+  }
+
+  return nullptr; // insufficient buffer
+}
+
+
+inline C10_HOST_DEVICE char* copy_args(
+    char* dst,
+    char* const end,
+    const char* arg) {
+  return copy_string(dst, end, arg);
+}
+
+#endif
+
+static inline C10_HOST_DEVICE char* copy_args(char* dst, char* const end) {
+  return dst;
+}
+
+template <typename T>
+C10_HOST_DEVICE char* copy_args(char* dst, char* const end, T arg) {
+  if (!dst || dst + C10_ASSERT_ARG_ALIGN_SIZE >= end) {
+    return nullptr;
+  }
+
+  // write length of arg
+  *reinterpret_cast<size_t*>(dst) = sizeof(T);
+  dst += C10_ASSERT_ARG_ALIGN_SIZE;
+
+  // write argument value
+  *reinterpret_cast<T*>(dst) = arg;
+  dst += C10_ASSERT_ARG_ALIGN_SIZE;
+
+  return dst;
+}
+
+template <typename T, typename... Args>
+C10_HOST_DEVICE char* copy_args(
+    char* dst,
+    char* const end,
+    T first,
+    Args... args) {
+  dst = copy_args(dst, end, first);
+  return copy_args(dst, end, args...);
+}
+
+template <typename... Args>
+C10_HOST_DEVICE __noinline__ void assert_fail(
+    CUDAAssert* assert_state,
+    AssertError error_code,
+    bool persistent,
+    AssertSource file,
+    uint32_t line,
+    AssertFormat format,
+    Args... args) {
+  assert(error_code != AssertError::OK);
+#ifdef __CUDA_ARCH__
+  if (atomicCAS(
+          const_cast<int32_t*>(&assert_state->error),
+          0,
+          static_cast<int32_t>(error_code)) != 0) {
+    return;
+  }
+#else
+  assert(false); // should never be called
+#endif
+
+  char* buffer = assert_state->buffer;
+  char* dst = buffer;
+  char* const end = dst + sizeof(assert_state->buffer);
+  dst = copy_args(dst, end, line, args...);
+
+  assert_state->persistent = persistent;
+  assert_state->file = file;
+  assert_state->line = line;
+  assert_state->format = format;
+  assert_state->length = dst ? dst - buffer : 0;
+}
+
+// handle case without format string, e.g. C10_KERNEL_ASSERT(false)
+static inline C10_HOST_DEVICE void assert_fail(
+    CUDAAssert* assert_state,
+    AssertError error_code,
+    bool persistent,
+    AssertSource file,
+    uint32_t line,
+    AssertFormat format) {
+  assert_fail(assert_state, error_code, persistent, file, line, format, 0);
+}
+
+static inline CUDAAssert* prepare_kernel_assert() {
+  auto current_stream = getCurrentCUDAStream();
+  return current_stream.assert_state();
+}
+
+} // namespace assert
+} // namespace native
+} // namespace at
+
+#define C10_PREPARE_KERNEL_ASSERT \
+  auto __c10_assert_state = at::native::assert::prepare_kernel_assert();
+
+#define C10_KERNEL_ASSERT(exp, format, ...) \
```

Comment:
These macros are *definitely* going to need some documentation about how to use them (now that they also assume that `__c10_assert_source` is defined :)